Feat/select: Select 컴포넌트 구현

se030 opened the pull request:
## 🤠 개요

<!--

- 이슈번호
- 한줄 설명

-->

- Closes #56 
- 현재 키보드 접근성을 제외한 기본 동작까지만 구현되어 있습니다.

## 💫 설명

<!--

- 현재 Pr 설명

-->

- 형태는 따로 정의하지 않았습니다.
 
  디자인이 포함된 요소들을 아직 구현하지 않아서 스크린샷은 현재 만들어져있는 컴포넌트들로만 구성했어요.

- select 태그와 관련된 로직은 useSelect 훅으로 분리했어요.

  작성된 옵션들을 select 태그 하위의 옵션으로 추가하고, 선택 시 select 여부를 변경해줍니다.

<br />

- 아래와 같은 이유로 린트 규칙을 끈 라인이 있습니다.
  <img width="738" alt="image" src="https://user-images.githubusercontent.com/63814960/226530816-316ec30e-3634-4fd8-9b42-5814902e6e74.png">

  - [피드백](https://github.com/c-h-w-h/cds/pull/65#pullrequestreview-1349785867) 받고 string으로 고정했습니다. 사실 `value ?? 'click here` 이게 쓰고싶어서 `string | null` 했나본데 초기값은 빈 문자열로 넣게 강제했어요. 유사한 동작은 삼항연산자로 작성할 수 있어요. documentation에 적어둘게요.

    ```ts
      <Select.Trigger>
        <Badge outline>{value ? value : 'click here'}</Badge>
      </Select.Trigger>
    ```

- 사실 키보드 접근성을 구현하려면 큰 문제가 하나 있는데, 옵션들이 없어졌다 생겼다(?) 하는 엘레먼트라서 ref 관리가 조금 까다로워요. (optionRefs 크게 수정해야 할듯..해서 여기까지만 리뷰받고 다른 브랜치 파서 작업할게요.)

## 📷 스크린샷


https://user-images.githubusercontent.com/63814960/226530228-5d398806-dddf-4014-8034-47722fb88978.mov

se030 committed:
feat: flexboxStyle default value 지정

se030 committed:
feat: Select 컴포넌트 기본 동작 구현

- 시각적인 형태를 제외한 기본 동작만 구현
- Trigger 요소는 Dropdown 컴포넌트 사용
- Option 요소를 클릭하면 선택된 값이 바뀌고 해당 선택이 HiddenSelect 내 option 태그에 반영

se030 committed:
feat: 옵션 선택했을 때 드롭다운 메뉴 닫기

se030 committed:
feat: Select 스토리 추가

se030 committed:
refactor: 커스텀훅 네이밍 변경

감춰진 select 태그와 관련된 동작을 모듈화

se030 committed:
fix: hiddenStyle 수정

se030 committed:
fix: option 태그가 이미 존재하는 경우 early return

prayinforrain commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,102 @@
+  overflow: hidden;
+  clip: rect(0 0 0 0);
+  clippath: inset(50%);
+  whitespace: nowrap;
```
`white-space` `clip-path`로 고쳐주세요 ㅎㅎ..

se030 replied:
푸시를 안했었는데 2f3d15a4281f245c40db6c84ac00f35be65cb528 에서 수정됐습니다!

prayinforrain commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,102 @@
+HiddenSelect.displayName = 'HiddenSelect';
+
+Select.Trigger = Dropdown.Trigger;
+Select.Options = Options;
+Select.Option = Option;
```
이거 구분할 수 있는 다른 이름 없을까요 ?ㅁ? 좀 묘해요

se030 replied:
OptionList 할까요 ..?

iyu88 replied:
OptionGroup...? 🤔

optgroup 태그의 역할과 헷갈릴 수도 있을 것 같긴하네요.
https://developer.mozilla.org/ko/docs/Web/HTML/Element/optgroup

se030 replied:
> optgroup 태그를 지원하기 위한 컴포넌트도 있으면 좋을 것 같아요!

우재님이 이렇게 말씀해주셔서 OptionGroup 말고 다른거 더 생각해볼게요 ㅋㅋㅋㅋ!!

prayinforrain replied:
사실 네임스페이스가 고정되어 있어서 그냥 List 해도 될 것 같아요.  
OptionList가 Option의 하위 컴포넌트같아 보이지 않을까 걱정되는데 문제 없다면 OptionList도 좋아요!

se030 replied:
딱히 좋은 네이밍이 생각 안나서 ㅜㅜ OptionList가 무난한 것 같아 그걸로 수정했습니다!

저 왜 코멘트 달 때마다 물음표가 생기죠 ...
<img width="313" alt="image" src="https://user-images.githubusercontent.com/63814960/226839862-a518c02d-56df-4125-8477-3190be8563ad.png">

prayinforrain commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,102 @@
+
+interface SelectContextInterface {
+  selectValue: (value: string) => void;
+  registerOption: ($el: HTMLLIElement, value: string) => void;
```
$ prefix가 무슨 의미인가요 ?ㅁ? 미리 확실히 해두면 컨벤션에 추가하고 쓸 수 있을 것 같아요

se030 replied:
이거 HTML 노드를 나타낼 때 붙이는 prefix로 알고있었는데 찾아보니까 뭐 공식적인 문서는 없고 유래는 jQuery래요.
개인적으로는 이렇게하면 변수 중에 뭐가 노드인지 구분이 잘 돼서 좋았는데 어떻게 정할까요 ?? 떼고 써도 괜찮아요.

iyu88 replied:
그러면 $가 가지는 노드의 의미는 살리면서

HTMLLIElement이니까

`$listItem` 제안해봅니다~

se030 replied:
컨펌 된건가요 ...ㅎㅎ 근데 지금까지 안붙이고 쓴 경우가 있을 것 같아서 저도 뗄까 싶어요 (?)

leesunmin1231 replied:
앗 이거 컨벤션으로 정해지면 저도 붙이고 써볼게욤
노드 구분이 잘되는 점은 좋은거 같아요

se030 replied:
> 그러면 $가 가지는 노드의 의미는 살리면서
> 
> HTMLLIElement이니까
> 
> `$listItem` 제안해봅니다~

```ts
    selectRef.current.childNodes.forEach(($option) => {
      if (!($option instanceof HTMLOptionElement)) return;
      $option.selected = $option === $target;
    });
```
저는 훅에서 이렇게 썼는데 **$ + 태그명 그대로** 쓰는거 어떨까요?

prayinforrain replied:
처음 답변 받고 집에 와서 계속 고민해봤는데 생각해보니 `useRef`로 사용하는 HTMLNode도 $를 써야할까 까지 생각하다가 잔것같아요 ㅋㅋ..  
childNodes, parent, getElementById같이 직접 접근하는 노드에 한해 세영님이 말씀하신대로 쓰는데 찬성이에요~ li 태그같은 축약어 이름도 풀어쓰면 좋겠지만 그게 가장 일반화해서 적용하기 쉬울 것 같아요.

iyu88 replied:
동의합니다 🙆‍♂️

se030 replied:
노드임을 드러내고 싶을 때 사용하면 좋을 것 같아요 ~ 지금 refCallback에 사용하는건 $ 뗄까요 ..?

prayinforrain reviewed (commented):
>Select 태그에서 선택되는 값을 외부에서 useState로 관리하고 싶은 경우도 있을 것 같은데, 어떤 타입을 사용하는지 알 수 없어서 any 사용했어요. unknown으로 하면 아래같은 타입 에러가 발생해요.

select 태그에서 value는 항상 string이지 않아요? 제가 생각하는 용도가 맞다면 string 타입으로 선언해도 될 것 같아요.

그리고 `optgroup` 태그를 지원하기 위한 컴포넌트도 있으면 좋을 것 같아요!  
사용자 입장에서 구현이 어려운 건 아니지만 코드의 통일성을 위해 ㅎㅎ..

iyu88 commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,99 @@
+  }
+  const { setIsOpen } = dropdownContext;
+
+  const onSelect = () => {
+    selectValue(value);
+    setIsOpen(false);
+  };
```
핸들러 이름이 간단하니까 좋네요 👍

iyu88 commented on src/components/Select/useSelect.tsx:
```diff
@@ -0,0 +1,42 @@
+    if (!selectRef.current) return;
+
+    const $target = selectRef.current.querySelector(`#${id}-${value}`);
+    [...selectRef.current.childNodes].forEach(($option) => {
+      if (!($option instanceof HTMLOptionElement)) return;
+      $option.selected = $option === $target;
+    });
```
spread 연산자를 사용하신 이유가 궁금해요~

se030 replied:
childNodes에서 반환되는 값이 유사배열객체인데 Array 메서드를 직접 사용할 수가 없어서 배열로 변환해서 사용했어요.

https://developer.mozilla.org/en-US/docs/Web/API/Node/childNodes
https://developer.mozilla.org/en-US/docs/Web/API/NodeList

iyu88 replied:
Note: Although NodeList is **not** an Array, it is **possible** to iterate over it with forEach(). 

NodeList 도 forEach로 가능하다는 의미로 해석했어요.
실제로 spread 없애도 에디터에서 빨간 줄 안뜨더라구요.
한 번 확인 부탁드릴게요!

se030 replied:
헉 그러네요 저 지금까지 항상 펼쳐서 사용했는데 ..ㅋㅋㅋㅋ🫠 떼버릴게요!

prayinforrain replied:
헐 저 이거보고 스프레드 안쓰면 안되지 않나..? 하고 속으로 생각했는데 충격받았어요

iyu88 commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,99 @@
+  return (
+    <li role="listitem" ref={refCallback} onClick={onSelect}>
+      {children}
+    </li>
```
`cursor: 'pointer'`
속성이 있으면 좋을 것 같아요!

스토리 확인하다보니까 Option을 누른건지 바탕을 누른건지 잘 모르겠더라구요!

iyu88 reviewed (commented):
커스텀 훅이 깔끔하네요 😊

leesunmin1231 commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,99 @@
+  useContext,
+} from 'react';
+
+import useSelect from './useSelect';
```
음 근데 얘는 hook이니까 따로 hook 디렉토리로 분류를 하는게 좋을까요?
그냥 사용하는 컴포넌트랑 같이 있는게 더 좋을 거 같기도 하고... 모르겠네요..

se030 replied:
움 저도 분리해야하나 잠깐 생각했는데 hook 디렉토리가 원래 없기도 했고, 어플리케이션 개발할 때의 비즈니스 로직 분리랑은 결이 달라서 컴포넌트랑 가깝게 두는게 낫겠다고 생각했어요. 훅끼리 모여있으면 오히려 어떤 훅이 어떤 컴포넌트에서 쓰이는건지 불분명할 것 같아서요 ..!!

se030 replied:
hooks 디렉토리 따로 분리 안하고 컴포넌트 폴더 내부에 두기로 결정함

leesunmin1231 commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,99 @@
+
+  const dropdownContext = useContext(DropdownContext);
+  if (!dropdownContext) {
+    throw new Error('Select component must have its own Dropdown context');
```
와 예외처리 좋아요

se030 replied:
있어야 하는 컨텍스트 없는 경우에는 에러 던지는게 좋을까요? (근데 사실은 발생할 수 없는 상황이에요)

se030 replied:
컴포넌트를 제대로 사용하면 발생하지 않는 상황이기 때문에 우선 Fragment 리턴하는걸로 결정함

se030 committed:
refactor: 리뷰 반영

- setValue 타입 문자열로 고정: .stories 용례 참고
- childNodes 스프레드 연산자 제거
- Option에 hover style 추가

se030 committed:
refactor: ChildrenProps 분리

속성 props 제외하고 children만 원하는 경우 사용할 수 있게 분리

prayinforrain commented on src/components/Select/index.tsx:
```diff
@@ -0,0 +1,107 @@
+      ref={refCallback}
+      onClick={onSelect}
+      css={css`
+        &:hover {
+          cursor: pointer;
+        }
```
이거 hover 의사 클래스 안넣어도 괜찮지 않나요?

se030 replied:
엥 전 바보네요

prayinforrain commented on src/utils/types/ChildrenProps.ts:
```diff
@@ -0,0 +1,5 @@
+import { ReactNode } from 'react';
+
+export interface ChildrenProps {
+  children: ReactNode;
+}
```
헐이거좋아요 제거 다 바꿔야겠어요

se030 replied:
근데 이건 관련 내용은 아니지만 ... 얘 친구도 고칠게 있는데 ㅜㅜ children이 없어야 하는 경우와 optional하게 있어도 되는 경우를 나눠서 사용해야해요.
```
import { ClassAttributes, HTMLAttributes, ReactNode } from 'react';

export type DefaultProps<T extends HTMLElement> = ClassAttributes<T> &
  HTMLAttributes<T> & {
    children?: ReactNode;
  };
```
지금 이렇게 되어있어서 Button, Image, Input 같은 애들 children 넣어도 에러 안뜨거든요 ...ㅎㅎ 별개로 하나 만들어줘야 할 것 같아요. 혹시 children?: 이 필요했던 컨텍스트가 뭐였는지 기억나시나요 ??

<img width="519" alt="image" src="https://user-images.githubusercontent.com/63814960/226832139-cc9d1bdc-e055-41f0-8783-43a3145e5ffa.png">

iyu88 replied:
그 때는 DefaultProps~ 인터페이스들에 css 가 있어서 

모든 컴포넌트 인터페이스에서 extends 하고 시작했는데

children 이 필요없는 컴포넌트는 extends 해서 사용하지 못하니까

만약을 대비해서 세영님이 

하나 만들어드릴게요~ 

하셨던거 같아요

이 컨텍스트 말씀하신게 맞겠죠?

https://github.com/c-h-w-h/cds/pull/19/files

se030 replied:
아 맞나봐요 ...!!!!!!! 그때는 emotion의 비밀을 깨닫기 전이라서 ...
지금 쓰이고있는 컴포넌트들이 Dropdown 제외하면 전부 children 필요하지 않은 것 같은데 그냥 지워도 되겠네요 ..ㅎㅎ

prayinforrain reviewed (approved):
운동 갔다 와서 한다고 했는데 운동가기 싫어하는 중이라 approve부터 찍고 갑니다~

se030 committed:
refactor: 리뷰 반영

- 노드 변수명 $li, $option 으로 수정
- context 존재하지 않는 경우 빈 Fragment 반환

se030 committed:
docs: 스토리 설명 일부 추가

se030 committed:
fix: DefaultProps에서 optionalChildren 제거

- 현재 스펙에 children이 옵셔널이어야 하는 케이스는 없고, 오히려 Input 등의 컴포넌트가 optional children을 가질 수 있는 케이스가 잘못되어서 제거
- 변경 사항에 따라 Dropdown 컴포넌트의 타입 일부 수정

se030 committed:
fix: 수정된 DropdownProps에 맞게 id, dropdownLabel 추가
